### PR TITLE
stdenv: introduce meta.supported

### PR DESCRIPTION
**Pull request by LnL7**
###### Motivation for this change

Fixes enableIfAvailable when config.allowBroken is set.

/cc @shlevy

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by LnL7** [comment]
Hrm, this shouldn't rebuild anything...

File: lib/meta.nix [hunk 1]
```diff
@@ -87,5 +87,5 @@ rec {
         else { parsed = elem; };
     in lib.matchAttrs pattern platform;
 
-  enableIfAvailable = p: if p.meta.available or true then [ p ] else [];
+  enableIfSupported = p: lib.optional p.meta.supported or true p;
```
Comment (shlevy) [inline lib/meta.nix]: I'm not sure if parens are literally required here but `lib.optional (p.meta.supported or true) p` seems much better IMO.

File: pkgs/stdenv/generic/make-derivation.nix [hunk 2]
```diff
@@ -241,6 +241,7 @@ rec {
                    && (if config.checkMetaRecursively or false
                        then lib.all (d: d.meta.available or true) references
                        else true);
+          supported = validity.supported or false;
```
Comment (shlevy) [inline pkgs/stdenv/generic/make-derivation.nix]: I think this does the wrong thing in case of unfree license.

**Review by shlevy — CHANGES_REQUESTED** [review]
I think really we need to move to a tristate model as @7c6f434c and @edolstra suggest: There's a difference between "this thing is not known to work on this platform" and "this thing is known not to work on this platform" and `allowBroken` shouldn't affect the latter.

File: pkgs/stdenv/generic/make-derivation.nix (hunk 2 above)
Comment (LnL7) [reply pkgs/stdenv/generic/make-derivation.nix]: Good point, should allowUnfree influence the enable condition or not?

**Comment by LnL7** [comment]
My main goal here is to fix the nix build, I'm fine with closing or reverting this if somebody implements the allowBroken changes.

**Comment by shlevy** [comment]
https://github.com/NixOS/nixpkgs/pull/38639

**Comment by LnL7** [comment]
Closing in favour of #38639.

**Pull request closed without merging**